### ISSUE TITLE
[GQSP] Precision issue in computing phase angles

### ISSUE THREAD

**Issue by anurudhp** [quantumlib/Qualtran#630]
Algorithm 1 of https://arxiv.org/abs/2308.01501: the QSP angles are calculated as $\theta_d = \arctan(|b_d / a_d|)$, $\phi_d = \arg(a_d / b_d)$.

1. Computing $\phi$ is numerically unstable when $a_d/b_d$ is zero, as small errors result in a non-zero angle.
2. Another issue is the edge case of $a_d = 0$ or $b_d = 0$, for which I could not figure out what the correct behaviour should be.

https://github.com/quantumlib/Qualtran/blob/642dec21cb6844bbe2f5f33e4312a4951f73e8c9/qualtran/bloqs/generalized_qsp.py#L190-L242

cc. @tanujkhattar

**Comment by tanujkhattar** [comment]
We can modify the current implementation of `phi[d] = np.angle(a) - np.angle(b)` to

```python
phi[d] = 0 if np.isclose(np.abs(b), 0) else np.angle(a * np.conj(b) / np.abs(b))
```

and the behavior should be well defined for the various cases you mentioned above? 

Although I'm not sure if this necessarily the problem following discussion on https://github.com/quantumlib/Qualtran/issues/631

**Comment by anurudhp** [comment]
fixed by #636
